File: benchmarks/suite5/matrix.c
#include <stdint.h>
#include <stdio.h>
#include "bench.h"

int main(void) {
  int64_t a[4][4];
  int64_t b[4][4];
  int64_t c[4][4];
  for (int i = 0; i < 4; i++) {
    for (int j = 0; j < 4; j++) {
      a[i][j] = (int64_t)(i + j);
      b[i][j] = (int64_t)(i * j + 1);
      c[i][j] = 0;
    }
  }
  const int64_t reps = 900000;
  int64_t trace = 0;
  for (int64_t r = 0; r < reps; r++) {
    for (int i = 0; i < 4; i++) {
      for (int j = 0; j < 4; j++) {
        int64_t s = 0;
        for (int k = 0; k < 4; k++) {
          s += a[i][k] * b[k][j];
        }
        c[i][j] = s;
      }
    }
    trace += c[r & 3][r & 3];
  }
  suite5_print_i64(trace);
  return 0;
}

File: benchmarks/suite5/matrix.zig
const bench = @import("bench.zig");
const c = @cImport({
    @cInclude("stdio.h");
});

pub fn main() void {
    var a: [4][4]i64 = undefined;
    var b: [4][4]i64 = undefined;
    var out: [4][4]i64 = undefined;
    for (0..4) |i| {
        for (0..4) |j| {
            a[i][j] = @intCast(i + j);
            b[i][j] = @intCast(i * j + 1);
            out[i][j] = 0;
        }
    }
    const reps: i64 = bench.opaqueI64(900000);
    var trace: i64 = 0;
    var r: i64 = 0;
    while (r < reps) : (r += 1) {
        for (0..4) |i| {
            for (0..4) |j| {
                var s: i64 = 0;
                for (0..4) |k| {
                    s += a[i][k] * b[k][j];
                }
                out[i][j] = s;
            }
        }
        trace += out[@intCast(r & 3)][@intCast(r & 3)];
    }
    bench.printI64(trace);
}
